### PR TITLE
Add Set-NsxEdgeFirewallRule

### LINKED ISSUE #565: Allow/Deny Edge Firewall
When will it be possible to change the configuration of an existing Firewall Rule in the Edge from Allow/Deny?

I can change the default rule but what about other rules?

Get-NsxEdge Edge01 | Get-NsxEdgeFirewall | Set-NsxEdgeFirewall -DefaultRuleAction deny

### PR DESCRIPTION
**Pull request by alagoutte**
Add Set-NsxEdgeFirewallRule and tests

Fix #565

### CODE REVIEW

**Comment by dcoghlan** [comment]
jenkins test this please

**Comment by alagoutte** [comment]
Hi Dale

Thanks for the jenkins test

it is possible to known why failed ?

File: tests/integration/04.Edge.Tests.ps1 [hunk 1]
```diff
@@ -545,17 +545,34 @@ Describe "Edge" {
...
-            $rule = Get-NsxEdge $name | Get-NsxEdgeFirewall | New-NsxEdgeFirewallRule -name "testrule19" -comment "testrule16" -action reject
+            $rule = Get-NsxEdge $name | Get-NsxEdgeFirewall | New-NsxEdgeFirewallRule -name "testrule19" -comment "testrule19" -action reject
             $rule | should not be $null
             $rule.action | should be "reject"
         }
 
+        It "Can modifiy an edge firewall rule" {
+            $rule = Get-NsxEdge $name | Get-NsxEdgeFirewall | New-NsxEdgeFirewallRule -name "testrule20" -comment "testrule20" -action accept
+            $rule | should not be $null
+            $rule.enabled | sould be $true
```
Comment (dcoghlan) [inline tests/integration/04.Edge.Tests.ps1]:
```suggestion
            $rule.enabled | should be $true
```

File: tests/integration/04.Edge.Tests.ps1 [hunk 2]
```diff
@@ -545,17 +545,34 @@ Describe "Edge" {
...
+            $rule.loggingEnabled | sould be $false
```
Comment (dcoghlan) [inline tests/integration/04.Edge.Tests.ps1]:
```suggestion
            $rule.loggingEnabled | should be $false
```

File: tests/integration/04.Edge.Tests.ps1 [hunk 3]
```diff
@@ -545,17 +545,34 @@ Describe "Edge" {
...
+            $rule.action | should be "accept"
+            $rule.name | should be "testrule20"
+            $rule.description | should be "testrule20"
+            $rule = $rule | Set-NsxEdgeFirewallRule -name "testrule21" -comment "testrule21" -loggingEnabled $true -enabled $false -action deny
+            $rule | should not be $null
+            $rule.enabled | sould be $false
```
Comment (dcoghlan) [inline tests/integration/04.Edge.Tests.ps1]:
```suggestion
            $rule.enabled | should be $false
```

File: tests/integration/04.Edge.Tests.ps1 [hunk 4]
```diff
@@ -545,17 +545,34 @@ Describe "Edge" {
...
+            $rule.loggingEnabled | sould be $true
```
Comment (dcoghlan) [inline tests/integration/04.Edge.Tests.ps1]:
```suggestion
            $rule.loggingEnabled | should be $true
```

**Comment by dcoghlan** [comment]
> Hi Dale
> 
> Thanks for the jenkins test
> 
> it is possible to known why failed ?

It was a spelling mistake in the tests. I've made some suggestions on where the spelling errors are. Once they are fixed, we will run it through the tests again and merge if it passes

**Comment by alagoutte** [comment]
Oups push the wrong commit

can you retest ?

**Comment by dcoghlan** [comment]
jenkins test this please

File: tests/integration/04.Edge.Tests.ps1 [hunk 5]
```diff
@@ -545,17 +545,34 @@ Describe "Edge" {
...
+            $rule = Get-NsxEdge $name | Get-NsxEdgeFirewall | New-NsxEdgeFirewallRule -name "testrule19" -comment "testrule19" -action reject
             $rule | should not be $null
             $rule.action | should be "reject"
         }
 
+        It "Can modifiy an edge firewall rule" {
+            $rule = Get-NsxEdge $name | Get-NsxEdgeFirewall | New-NsxEdgeFirewallRule -name "testrule20" -comment "testrule20" -action accept
+            $rule | should not be $null
+            $rule.enabled | should be $true
+            $rule.loggingEnabled | should be $false
```
Comment (dcoghlan) [inline tests/integration/04.Edge.Tests.ps1]:
```suggestion
            $rule.loggingEnabled | should be "false"
```
LoggingEnabled is actually a string

File: tests/integration/04.Edge.Tests.ps1 [hunk 6]
```diff
@@ -545,17 +545,34 @@ Describe "Edge" {
...
+            $rule.action | should be "accept"
+            $rule.name | should be "testrule20"
+            $rule.description | should be "testrule20"
+            $rule = $rule | Set-NsxEdgeFirewallRule -name "testrule21" -comment "testrule21" -loggingEnabled $true -enabled $false -action deny
+            $rule | should not be $null
+            $rule.enabled | should be $false
+            $rule.loggingEnabled | should be $true
```
Comment (dcoghlan) [inline tests/integration/04.Edge.Tests.ps1]:
```suggestion
            $rule.loggingEnabled | should be "true"
```
loggingEnabled is actually a string

**Comment by dcoghlan** [comment]
Tests failed with the following error
```
MESSAGE:
Expected $false, but got 'false'.
+++++++++++++++++++
STACK TRACE:
at <ScriptBlock>, /var/jenkins/workspace/CorePowerNsx641/tests/integration/04.Edge.Tests.ps1: line 563
563:             $rule.loggingEnabled | should be $false
```

**Comment by alagoutte** [comment]
Thanks! 

can you relaunch test ?

**Comment by dcoghlan** [comment]
jenkins test this please

**Comment by dcoghlan** [comment]
Test failed due to the function not being exported. Please add the function to /module/include.ps1
```
MESSAGE:
The term 'Set-NsxEdgeFirewallRule' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
+++++++++++++++++++
STACK TRACE:
at <ScriptBlock>, C:\jenkins\workspace\WindowsPowerNsx641\tests\integration\04.Edge.Tests.ps1: line 567
567:             $rule = $rule | Set-NsxEdgeFirewallRule -name "testrule21" -comment "testrule21" -loggingEnabled $true -enabled $false -action deny```

**Comment by alagoutte** [comment]
Sorry, Add Set-NsxEdgeFirewallRule to module/Include.ps1

**Comment by dcoghlan** [comment]
jenkins test this please

File: tests/integration/04.Edge.Tests.ps1 [hunk 7]
```diff
@@ -545,17 +545,34 @@ Describe "Edge" {
...
-            $rule = Get-NsxEdge $name | Get-NsxEdgeFirewall | New-NsxEdgeFirewallRule -name "testrule19" -comment "testrule16" -action reject
+            $rule = Get-NsxEdge $name | Get-NsxEdgeFirewall | New-NsxEdgeFirewallRule -name "testrule19" -comment "testrule19" -action reject
             $rule | should not be $null
             $rule.action | should be "reject"
         }
 
+        It "Can modifiy an edge firewall rule" {
+            $rule = Get-NsxEdge $name | Get-NsxEdgeFirewall | New-NsxEdgeFirewallRule -name "testrule20" -comment "testrule20" -action accept
+            $rule | should not be $null
+            $rule.enabled | should be $true
```
Comment (dcoghlan) [inline tests/integration/04.Edge.Tests.ps1]:
```suggestion
            $rule.enabled | should be "true"
```

File: tests/integration/04.Edge.Tests.ps1 [hunk 8]
```diff
@@ -545,17 +545,34 @@ Describe "Edge" {
...
+            $rule.loggingEnabled | should be "false"
+            $rule.action | should be "accept"
+            $rule.name | should be "testrule20"
+            $rule.description | should be "testrule20"
+            $rule = $rule | Set-NsxEdgeFirewallRule -name "testrule21" -comment "testrule21" -loggingEnabled $true -enabled $false -action deny
+            $rule | should not be $null
+            $rule.enabled | should be $false
```
Comment (dcoghlan) [inline tests/integration/04.Edge.Tests.ps1]:
```suggestion
            $rule.enabled | should be "false"
```

**Comment by dcoghlan** [comment]
Please update test to check for $rule.enabled as a string

```[-] Can modifiy an edge firewall rule 5.05s
        Expected: {False}
        But was:  {false}
        at <ScriptBlock>, C:\jenkins\workspace\WindowsPowerNsx641\tests\integration\04.Edge.Tests.ps1: line 569
        569:             $rule.enabled | should be $false```

**Comment by alagoutte** [comment]
Done

**Comment by dcoghlan** [comment]
jenkins test this please

**Comment by Mooihoek** [comment]
I just did Update-Module PowerNSX but receive the following error: 
 Set-NsxEdgeFirewallRule : The term 'Set-NsxEdgeFirewallRule' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.

Am I missing something? Thought this new command had been included?

**Comment by alagoutte** [comment]
Hi @Mooihoek 

The patch is merged but there is now yet updated release of module on psgallery

You need to manually download master branch (or clone)

**Comment by Mooihoek** [comment]
> Hi @Mooihoek
> 
> The patch is merged but there is now yet updated release of module on psgallery
> 
> You need to manually download master branch (or clone)

great thanks so much!

**Pull request merged**